File: db/schemas/stremio.py
"""Stremio addon schemas for catalog and stream responses."""

from typing import Literal

from pydantic import BaseModel, Field, model_validator


class PosterData(BaseModel):
    """Data required for poster generation."""

    id: str
    poster: str
    title: str
    imdb_rating: float | None = None
    is_add_title_to_poster: bool = False


class Catalog(BaseModel):
    """Stremio catalog definition."""

    id: str
    name: str
    type: str


class Video(BaseModel):
    """Stremio video (episode) definition."""

    id: str
    title: str
    released: str | None = None
    season: int | None = None
    episode: int | None = None
    thumbnail: str | None = None


class Meta(BaseModel):
    """Stremio metadata object."""

    id: str = Field(alias="_id")
    name: str = Field(alias="title")
    type: str = Field(default="movie")
    poster: str | None = None
    background: str | None = None
    videos: list[Video] | None = None
    country: str | None = None
    language: str | None = Field(None, alias="tv_language")
    logo: str | None = None
    genres: list[str] | None = None
    description: str | None = None
    runtime: str | None = None
    website: str | None = None
    imdbRating: str | float | None = Field(None, alias="imdb_rating")
    releaseInfo: str | int | None = Field(None, alias="year")
    cast: list[str] | None = Field(None, alias="stars")

    @model_validator(mode="after")
    def parse_meta(self) -> "Meta":
        if self.releaseInfo:
            self.releaseInfo = f"{self.releaseInfo}-" if self.type == "series" else str(self.releaseInfo)
        if self.imdbRating:
            self.imdbRating = str(self.imdbRating)
        return self


class MetaItem(BaseModel):
    """Wrapper for a single meta object."""

    meta: Meta


class Metas(BaseModel):
    """Collection of meta objects."""

    metas: list[Meta] = Field(default_factory=list)


class StreamBehaviorHints(BaseModel):
    """Stremio stream behavior hints."""

    notWebReady: bool | None = None
    bingeGroup: str | None = None
    proxyHeaders: dict[Literal["request", "response"], dict] | None = None
    filename: str | None = None
    videoSize: int | None = None


class Stream(BaseModel):
    """Stremio stream object."""

    name: str
    description: str
    infoHash: str | None = None
    fileIdx: int | None = None
    url: str | None = None
    nzbUrl: str | None = None  # Direct NZB URL for Stremio v5 native NNTP streaming
    ytId: str | None = None
    externalUrl: str | None = None
    behaviorHints: StreamBehaviorHints | None = None
    sources: list[str] | None = None


class Streams(BaseModel):
    """Collection of streams."""

    streams: list[Stream] | None = Field(default_factory=list)


class RichStreamMetadata(BaseModel):
    """Rich metadata for frontend display - supplements the Stremio Stream format."""

    id: str  # stream_id - used as stream ID for voting/editing
    info_hash: str
    name: str  # Stream display name (torrent title)

    # Quality attributes (normalized from separate tables)
    resolution: str | None = None  # 4k, 1080p, 720p, 480p
    quality: str | None = None  # web-dl, bluray, cam, hdtv
    codec: str | None = None  # x264, x265, hevc, av1
    bit_depth: str | None = None  # 8bit, 10bit, 12bit

    # Multi-value quality attributes (from normalized tables)
    audio_formats: list[str] = Field(default_factory=list)  # AAC, DTS, Atmos, TrueHD
    channels: list[str] = Field(default_factory=list)  # 2.0, 5.1, 7.1
    hdr_formats: list[str] = Field(default_factory=list)  # HDR10, HDR10+, Dolby Vision
    languages: list[str] = Field(default_factory=list)

    # Release flags
    is_remastered: bool = False
    is_upscaled: bool = False
    is_proper: bool = False
    is_repack: bool = False
    is_extended: bool = False
    is_complete: bool = False
    is_dubbed: bool = False
    is_subbed: bool = False

    # Stream info
    source: str
    size: int | None = None  # size in bytes
    size_display: str | None = None  # formatted size (e.g., "2.5 GB")
    seeders: int | None = None
    uploader: str | None = None
    uploaded_at: str | None = None
    cached: bool = False


class RichStream(BaseModel):
    """Combined stream data for both Stremio addon and frontend API."""

    # Stremio-compatible stream (for addon endpoint)
    stream: Stream
    # Rich metadata for frontend (for catalog API)
    metadata: RichStreamMetadata

File: backend/src/models/stremio.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

// ─── Manifest ─────────────────────────────────────────────────────────────────

#[derive(Serialize)]
pub struct Manifest {
    pub id: String,
    pub version: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none", rename = "contactEmail")]
    pub contact_email: Option<String>,
    pub description: String,
    pub logo: String,
    #[serde(rename = "behaviorHints")]
    pub behavior_hints: ManifestBehaviorHints,
    /// Mixed array: "catalog" (plain string) + typed resource objects.
    pub resources: Vec<Value>,
    pub types: Vec<String>,
    pub catalogs: Vec<CatalogDef>,
}

#[derive(Serialize)]
pub struct ManifestBehaviorHints {
    pub configurable: bool,
    #[serde(rename = "configurationRequired")]
    pub configuration_required: bool,
}

#[derive(Serialize)]
pub struct CatalogDef {
    pub id: String,
    #[serde(rename = "type")]
    pub catalog_type: String,
    pub name: String,
    pub extra: Vec<ExtraField>,
}

#[derive(Serialize)]
pub struct ExtraField {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none", rename = "isRequired")]
    pub is_required: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub options: Option<Vec<String>>,
}

// ─── Catalog/Metas ────────────────────────────────────────────────────────────

#[derive(Serialize, Deserialize)]
pub struct Metas {
    pub metas: Vec<MetaPreview>,
}

#[derive(Serialize, Deserialize)]
pub struct MetaPreview {
    pub id: String,
    #[serde(rename = "type")]
    pub media_type: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none", rename = "releaseInfo")]
    pub release_info: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub poster: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub background: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

// ─── Meta ─────────────────────────────────────────────────────────────────────

#[derive(Serialize)]
pub struct MetaItem {
    pub meta: Meta,
}

#[derive(Serialize)]
pub struct Meta {
    pub id: String,
    #[serde(rename = "type")]
    pub media_type: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none", rename = "releaseInfo")]
    pub release_info: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub poster: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub background: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub logo: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub runtime: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub website: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub language: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub country: Option<String>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub genres: Vec<String>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub cast: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "imdbRating")]
    pub imdb_rating: Option<String>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub videos: Vec<Video>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub links: Option<Vec<Value>>,
}

#[derive(Serialize)]
pub struct Video {
    pub id: String,
    pub title: String,
    pub released: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub overview: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thumbnail: Option<String>,
    pub season: i32,
    pub episode: i32,
}
